lib/Segments/WooCommerce.php

```diff
@@ -146,22 +167,35 @@ public function synchronizeGuestCustomer($orderId) {
     }
   }
 
-  public function synchronizeCustomers() {
+  public function synchronizeCustomers(int $countOfSynchronized = 0): int {
+    if ($countOfSynchronized === 0) {
+      $this->resetSynchronization();
+    }
+
     $this->wpSegment->synchronizeUsers(); // synchronize registered users
 
     $this->markRegisteredCustomers();
+
     $insertedUsersEmails = $this->insertSubscribersFromOrders();
-    $this->removeUpdatedSubscribersWithInvalidEmail($insertedUsersEmails);
-    unset($insertedUsersEmails);
-    $this->updateFirstNames();
-    $this->updateLastNames();
-    $this->insertUsersToSegment();
-    $this->unsubscribeUsersFromSegment();
-    $this->removeOrphanedSubscribers();
-    $this->updateStatus();
-    $this->updateGlobalStatus();
+    $this->updateNames($insertedUsersEmails);
+
+    if (count($insertedUsersEmails) < self::BATCH_SIZE) {
+      $this->insertUsersToSegment();
+      $this->unsubscribeUsersFromSegment();
+      $this->removeOrphanedSubscribers();
+      $this->updateStatus();
+      $this->updateGlobalStatus();
+    }
 
-    return true;
+    return count($insertedUsersEmails);
+  }
+
+  public function resetSynchronization(): void {
+    $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+    $this->connection->executeQuery("
+      UPDATE {$subscribersTable}
```

Comment:
This updates all rows in subscribers table. I'm a bit afraid that this may cause some queuing of queries because of locking.